src/proof/PGTransformationAlgorithms.cc

```diff
@@ -555,73 +555,64 @@ void ProofGraph::proofTransformAndRestructure(const double left_time, const int
-
-                    // NOTE not clear how this might happen
-                    assert(n->getAnt1() != n->getAnt2());
-                    if (n->getAnt1() == n->getAnt2()) {
-                        opensmt_error("Bad proof structure!");
-                    }
+                    assert(n->getAnt2());
+                    assert((n->getAnt1()->getAnt1() and n->getAnt1()->getAnt2())
+                           or (n->getAnt1()->getAnt1() == nullptr and n->getAnt1()->getAnt2() == nullptr));
+                    assert((n->getAnt2()->getAnt1() and n->getAnt2()->getAnt2())
+                           or (n->getAnt2()->getAnt1() == nullptr && n->getAnt2()->getAnt2() == nullptr));
+
                     //Look for pivot in antecedents
                     short f1 = n->getAnt1()->hasOccurrenceBin(n->getPivot());
                     bool piv_in_ant1 = (f1 != -1);
                     short f2 = n->getAnt2()->hasOccurrenceBin(n->getPivot());
                     bool piv_in_ant2 = (f2 != -1);
-                    assert(!(f1 == 1 && f2 == 1) && !(f1 == 0 && f2 == 0));
-                    //Easy case: pivot still in both antecedents
-                    //Sufficient to propagate modifications via merge
-                    if (piv_in_ant1 && piv_in_ant2) {
-                        for (clauseid_t clauseid : n->getResolvents()) {
-                            if (getNode(clauseid) != nullptr) {
-                                //assert(!isSetVisited1(clauseid));
-                                q.push_back(clauseid);
-                            }
+                    assert(not(f1 == 1 and f2 == 1) and not(f1 == 0 and f2 == 0));
+                    if (piv_in_ant1 and piv_in_ant2) {
+                        //Easy case: pivot still in both antecedents
+                        //Sufficient to propagate modifications via merge
+                        for (clauseid_t resolvent : n->getResolvents()) {
+                            if (getNode(resolvent)) { q.push_back(resolvent); }
                         }
 
-                        if (isSetVisited2(n->getAnt1()->getId()) || isSetVisited2(n->getAnt2()->getId())) {
-                            mergeClauses(n->getAnt1()->getClause(), n->getAnt2()->getClause(), n->getClause(),
-                                         n->getPivot());
+                        if (isSetVisited2(n->getAnt1()->getId()) or isSetVisited2(n->getAnt2()->getId())) {
+                            mergeClauses(n->getAnt1()->getClause(), n->getAnt2()->getClause(), n->getClause(), n->getPivot());
                             setVisited2(id);
                         }
                         // NOTE extra check
                         if (proofCheck() > 1) { checkClause(n->getId()); }
 
                         if (do_transf) {
+                            RuleContext ra1, ra2;
                             //Look for rules applicability
                             getRuleContext(n->getAnt1()->getId(), n->getId(), ra1);
```

Comment:
The return value of these are not checked.  It looks like changing the signature to `RuleContext ProofGraph::getRuleContext(clauseid_t idv3, clauseid_t idv)`.  Do you think that'd be better?

Reply:
Signature changed as you proposed.